dhall/src/Dhall/TypeCheck.hs

```diff
@@ -937,86 +937,85 @@ infer typer = loop
                 then return ()
                 else die (UnusedHandler diffT)
 
-            (my₀, _T₁') <- do
-                case mT₁ of
-                    Just _T₁ -> do
-                        return (Nothing, eval values _T₁)
-
-                    Nothing -> do
-                        case Dhall.Map.uncons yTs' of
-                            Nothing -> do
-                                die MissingMergeType
-
-                            Just (y₀, _T', _) -> do
-                                _T₁' <- do
-                                    case Dhall.Map.lookup y₀ yUs' of
-                                        Nothing -> do
-                                            die (UnusedHandler diffU)
+            if Data.Set.null diffU
+                then return ()
+                else die (MissingHandler diffU)
 
-                                        Just Nothing -> do
-                                            return _T'
+            let match _y _T₀' Nothing =
+                    return _T₀'
 
-                                        Just (Just _) -> do
-                                            case Eval.toVHPi _T' of
-                                                Just (x, _A₀', _T₀') -> do
-                                                    return (_T₀' (fresh ctx x))
+                match y handler' (Just _A₁') =
+                    case Eval.toVHPi handler' of
+                        Just (x, _A₀', _T₀') -> do
+                            if Eval.conv values _A₀' _A₁'
+                                then do
+                                    let _T₁' = _T₀' (fresh ctx x)
 
-                                                Nothing -> do
-                                                    let _T'' = quote names _T'
+                                    let _T₁'' = quote names _T₁'
 
-                                                    die (HandlerNotAFunction y₀ _T'')
+                                    _ <- loop ctx _T₁''
```

Comment:
I think the free variable check here can be done more efficiently, but I haven't yet figured out how…